agent/testflinger_agent/agent.py

```diff
@@ -143,11 +153,34 @@ def unpack_attachments(self, job_data: dict, cwd: Path):
         with tempfile.NamedTemporaryFile(suffix="tar.gz") as archive_tmp:
             archive_path = Path(archive_tmp.name)
             # download attachment archive
+            logger.info(f"Downloading attachments for {job_id}")
             self.client.get_attachments(job_id, path=archive_path)
             # extract archive into the attachments folder
+            logger.info(f"Unpacking attachments for {job_id}")
             with tarfile.open(archive_path, "r:gz") as tar:
                 tar.extractall(cwd / ATTACHMENTS_DIR, filter=secure_filter)
 
+        # side effect: remove all attachment data from `job_data`
+        # (so there is no interference with existing processes, especially
+        # provisioning or firmware update, which are triggered when these
+        # sections are not empty)
+        for phase in ("provision", "firmware_update", "test"):
+            phase_str = f"{phase}_data"
+            try:
+                phase_data = job_data[phase_str]
+            except KeyError:
+                pass
+            else:
+                try:
+                    del phase_data["attachments"]
+                except KeyError:
+                    pass
+                else:
+                    # it may be the case that attachments were the only data
+                    # included for this phase, so the phase can now be removed
+                    if not phase_data:
+                        del job_data[phase_str]
+
```

Comment:
Is it possible to simplify this logic? We can try something like this:

```suggestion
        for phase in ("provision", "firmware_update", "test"):
            phase_str = f"{phase}_data"
            phase_data = job_data.get(phase_str)

            if phase_data:
                # Attempt to delete attachments if they exist
                phase_data.pop("attachments", None)

                # If phase_data is empty, remove it from job_data
                if not phase_data:
                    del job_data[phase_str]
```

Reply:
I think I want to keep the outer try-except but I admit that the nested one makes the logic harder to follow (Thomas made the very same remark as well). Using `pop` is a great idea to resolve that and I'll definitely follow that suggestion.